# Segmenting & Clustering Neighborhoods in Toronto

In [1]:
import pandas as pd
import numpy as np
import requests
from bs4 import BeautifulSoup

from geopy.geocoders import Nominatim
!conda install -c conda-forge folium=0.5.0
import folium

from sklearn.cluster import KMeans

import matplotlib.cm as cm
import matplotlib.colors as colors

Solving environment: ...working... done

# All requested packages already installed.



## Part-1: Import the data

In [2]:
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
html_data  = requests.get(url).text

In [3]:
soup = BeautifulSoup(html_data,"html.parser")
soup.find_all('title')

[<title>List of postal codes of Canada: M - Wikipedia</title>]

In [4]:
table_contents = pd.DataFrame(columns=['Postal Code', 'Borough', 'Neighborhood'])

table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':  #To skip if the borough name is 'Not Assigned'
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

### Visualize the data

In [5]:
df=pd.DataFrame(table_contents)
df

PostalCode                                            Borough  \
0          M3A                                         North York   
1          M4A                                         North York   
2          M5A                                   Downtown Toronto   
3          M6A                                         North York   
4          M7A                                       Queen's Park   
..         ...                                                ...   
98         M8X                                          Etobicoke   
99         M4Y                                   Downtown Toronto   
100        M7Y  East TorontoBusiness reply mail Processing Cen...   
101        M8Y                                          Etobicoke   
102        M8Z                                          Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

In [6]:
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood  
0                                            Parkwoods  
1                                     Victoria Village  
2                            Regent Park, Harbourfront  
3                     Lawrence Manor, Lawrence Heights  
4                        Ontario Provincial Government  
..                                                 ...  
98       The Kingsway, Montgomery Road, Old Mill North  
99                                Church and Wellesley  
100                                     Enclave of M4L  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  
102  Mimico NW, The Queensway West, South of Bloor,...  

[103 rows x 3 columns]

### Checking any record where neighborhood is "Not assigned"

In [7]:
df.Neighborhood.str.count("Not assigned").sum()

0

In [8]:
df.shape

(103, 3)

### Above result shows that we have rows = 103 and columns =3

## Part-2: Latitude & Longitude of each neighborhood

In [9]:
data = pd.read_csv("https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv")
data

Postal Code   Latitude  Longitude
0           M1B  43.806686 -79.194353
1           M1C  43.784535 -79.160497
2           M1E  43.763573 -79.188711
3           M1G  43.770992 -79.216917
4           M1H  43.773136 -79.239476
..          ...        ...        ...
98          M9N  43.706876 -79.518188
99          M9P  43.696319 -79.532242
100         M9R  43.688905 -79.554724
101         M9V  43.739416 -79.588437
102         M9W  43.706748 -79.594054

[103 rows x 3 columns]

### Since the dimensions of both dataframes are the same, we can join on postal code

In [10]:
combined_data = df.join(data.set_index('Postal Code'), on='PostalCode', how='inner')
combined_data

PostalCode                Borough  \
0          M3A             North York   
1          M4A             North York   
2          M5A       Downtown Toronto   
3          M6A             North York   
4          M7A           Queen's Park   
..         ...                    ...   
98         M8X              Etobicoke   
99         M4Y       Downtown Toronto   
100        M7Y  East Toronto Business   
101        M8Y              Etobicoke   
102        M8Z              Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                            Regent Park, Harbourfront  43.654260 -79.360636  
3                     Lawrence Manor, Lawrence Heights  43.718518 -79.464763  
4                        Ontario Provincial Government  43.662301 -79.389494  
..                                                 ...        ...        ...  
98       The Kingsway, Montgomery Road, Old Mill North  43.653654 -79.506944  
99                                Church and Wellesley  43.665860 -79.383160  
100                                     Enclave of M4L  43.662744 -79.321558  
101  Old Mill South, King's Mill Park, Sunnylea, Hu...  43.636258 -79.498509  
102  Mimico NW, The Queensway West, South of Bloor,...  43.628841 -79.520999  

[103 rows x 5 columns]

### Shape of combined data

In [11]:
combined_data.shape

(103, 5)

### Above result shows that the combined data has 103 rows and 5 columns

## Part-3: Use geopy library to get the latitude and longitude values of Toronto

In [12]:
address = 'Toronto, Ontario'

geolocator = Nominatim(user_agent="toronto_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The coordinates of Toronto are {}, {}.'.format(latitude, longitude))

The coordinates of Toronto are 43.6534817, -79.3839347.


### Create a map of Toronto with neighborhoods superimposed on top

In [13]:
map_Toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# adding markers to map
for latitude, longitude, borough, neighborhood in zip(combined_data['Latitude'], combined_data['Longitude'], combined_data['Borough'], combined_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [latitude, longitude],
        radius=5,
        popup=label,
        color='red',
        fill=True
        ).add_to(map_Toronto)  
    
map_Toronto

### Define Foursquare Credentials and Version

In [14]:
CLIENT_ID = 'ZJRX4RJNBQKRMU1FQE4UEDX2E1XZXPGJ4ISGI3DORE4CSLJ2' 
CLIENT_SECRET = '2ZJHTMITKP1COXXG2SNI5I5AOG0WMVXELF3ENHW2OCMNYXGS'
VERSION = '20180605' # Foursquare API version
LIMIT = 100

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: ZJRX4RJNBQKRMU1FQE4UEDX2E1XZXPGJ4ISGI3DORE4CSLJ2
CLIENT_SECRET:2ZJHTMITKP1COXXG2SNI5I5AOG0WMVXELF3ENHW2OCMNYXGS


### Let's create a function to get all the venue types in Toronto

In [15]:
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

### Retrieve all the venues in Toronto for each neighborhood

In [16]:
venues_in_toronto = getNearbyVenues(combined_data['Neighborhood'], combined_data['Latitude'], combined_data['Longitude'])

Parkwoods
Victoria Village
Regent Park, Harbourfront
Lawrence Manor, Lawrence Heights
Ontario Provincial Government
Islington Avenue
Malvern, Rouge
Don Mills North
Parkview Hill, Woodbine Gardens
Garden District, Ryerson
Glencairn
West Deane Park, Princess Gardens, Martin Grove, Islington, Cloverdale
Rouge Hill, Port Union, Highland Creek
Don Mills South
Woodbine Heights
St. James Town
Humewood-Cedarvale
Eringate, Bloordale Gardens, Old Burnhamthorpe, Markland Wood
Guildwood, Morningside, West Hill
The Beaches
Berczy Park
Caledonia-Fairbanks
Woburn
Leaside
Central Bay Street
Christie
Cedarbrae
Hillcrest Village
Bathurst Manor, Wilson Heights, Downsview North
Thorncliffe Park
Richmond, Adelaide, King
Dufferin, Dovercourt Village
Scarborough Village
Fairview, Henry Farm, Oriole
Northwood Park, York University
The Danforth  East
Harbourfront East, Union Station, Toronto Islands
Little Portugal, Trinity
Kennedy Park, Ionview, East Birchmount Park
Bayview Village
Downsview East
The Danforth

In [17]:
venues_in_toronto.shape

(1992, 7)

### Above result shows that We have 1992 rows and 7 columns

### Now Check the sample data

In [18]:
venues_in_toronto.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0         Parkwoods              43.753259              -79.329656   
1         Parkwoods              43.753259              -79.329656   
2         Parkwoods              43.753259              -79.329656   
3  Victoria Village              43.725882              -79.315572   
4  Victoria Village              43.725882              -79.315572   

                    Venue  Venue Latitude  Venue Longitude  \
0                     KFC       43.754387       -79.333021   
1         Brookbanks Park       43.751976       -79.332140   
2           Variety Store       43.751974       -79.333114   
3  Victoria Village Arena       43.723481       -79.315635   
4               Portugril       43.725819       -79.312785   

          Venue Category  
0   Fast Food Restaurant  
1                   Park  
2      Food & Drink Shop  
3           Hockey Arena  
4  Portuguese Restaurant

### Check how many venues were returned for each neighborhood for Toronto

In [19]:
venues_in_toronto.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                             
Agincourt                                                            4   
Alderwood, Long Branch                                               7   
Bathurst Manor, Wilson Heights, Downsview North                     17   
Bayview Village                                                      4   
Bedford Park, Lawrence Manor East                                   23   
...                                                                ...   
Willowdale, Newtonbrook                                              1   
Woburn                                                               3   
Woodbine Heights                                                     4   
York Mills West                                                      3   
York Mills, Silver Hills                                             1   

                                                 Neighborhood Longitude  \
Neighborhood                                                              
Agincourt                                                             4   
Alderwood, Long Branch                                                7   
Bathurst Manor, Wilson Heights, Downsview North                      17   
Bayview Village                                                       4   
Bedford Park, Lawrence Manor East                                    23   
...                                                                 ...   
Willowdale, Newtonbrook                                               1   
Woburn                                                                3   
Woodbine Heights                                                      4   
York Mills West                                                       3   
York Mills, Silver Hills                                              1   

                                                 Venue  Venue Latitude  \
Neighborhood                                                             
Agincourt                                            4               4   
Alderwood, Long Branch                               7               7   
Bathurst Manor, Wilson Heights, Downsview North     17              17   
Bayview Village                                      4               4   
Bedford Park, Lawrence Manor East                   23              23   
...                                                ...             ...   
Willowdale, Newtonbrook                              1               1   
Woburn                                               3               3   
Woodbine Heights                                     4               4   
York Mills West                                      3               3   
York Mills, Silver Hills                             1               1   

                                                 Venue Longitude  \
Neighborhood                                                       
Agincourt                                                      4   
Alderwood, Long Branch                                         7   
Bathurst Manor, Wilson Heights, Downsview North               17   
Bayview Village                                                4   
Bedford Park, Lawrence Manor East                             23   
...                                                          ...   
Willowdale, Newtonbrook                                        1   
Woburn                                                         3   
Woodbine Heights                                               4   
York Mills West                                                3   
York Mills, Silver Hills                                       1   

                                                 Venue Category  
Neighborhood                                                     
Agincourt                                                     4  
Alderwood, Long Branch                                        7 

### Check how many unique categories can be curated from all the returned venues

In [20]:
print('There are {} uniques categories.'.format(len(venues_in_toronto['Venue Category'].unique())))

There are 257 uniques categories.


### Now analyze each neighborhood

In [21]:
# one hot encoding
Toronto_onehot = pd.get_dummies(venues_in_toronto[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
Toronto_onehot['Neighborhood'] = venues_in_toronto['Neighborhood'] 

# move neighborhood column to the first column
fixed_columns = [Toronto_onehot.columns[-1]] + list(Toronto_onehot.columns[:-1])
Toronto_onehot = Toronto_onehot[fixed_columns]

Toronto_onehot.head()

Yoga Studio  Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0            0                  0               0                  0        0   
1            0                  0               0                  0        0   
2            0                  0               0                  0        0   
3            0                  0               0                  0        0   
4            0                  0               0                  0        0   

   Airport Gate  Airport Lounge  Airport Service  Airport Terminal  \
0             0               0                0                 0   
1             0               0                0                 0   
2             0               0                0                 0   
3             0               0                0                 0   
4             0               0                0                 0   

   American Restaurant  ...  Trail  Train Station  \
0                    0  ...      0              0   
1                    0  ...      0              0   
2                    0  ...      0              0   
3                    0  ...      0              0   
4                    0  ...      0              0   

   Vegetarian / Vegan Restaurant  Video Game Store  Vietnamese Restaurant  \
0                              0                 0                      0   
1                              0                 0                      0   
2                              0                 0                      0   
3                              0                 0                      0   
4                              0                 0                      0   

   Warehouse Store  Wine Bar  Wine Shop  Wings Joint  Women's Store  
0                0         0          0            0              0  
1                0         0          0            0              0  
2                0         0          0            0              0  
3                0         0          0            0              0  
4                0         0          0            0              0  

[5 rows x 257 columns]

In [22]:
Toronto_onehot.shape

(1992, 257)

### We can group rows by neighborhood and taking the mean of the frequency of occurrence of each category

In [23]:
Toronto_grouped = Toronto_onehot.groupby('Neighborhood').mean().reset_index()
Toronto_grouped

Neighborhood  Yoga Studio  \
0                                          Agincourt          0.0   
1                             Alderwood, Long Branch          0.0   
2    Bathurst Manor, Wilson Heights, Downsview North          0.0   
3                                    Bayview Village          0.0   
4                  Bedford Park, Lawrence Manor East          0.0   
..                                               ...          ...   
96                           Willowdale, Newtonbrook          0.0   
97                                            Woburn          0.0   
98                                  Woodbine Heights          0.0   
99                                   York Mills West          0.0   
100                         York Mills, Silver Hills          0.0   

     Accessories Store  Adult Boutique  Afghan Restaurant  Airport  \
0                  0.0             0.0                0.0      0.0   
1                  0.0             0.0                0.0      0.0   
2                  0.0             0.0                0.0      0.0   
3                  0.0             0.0                0.0      0.0   
4                  0.0             0.0                0.0      0.0   
..                 ...             ...                ...      ...   
96                 0.0             0.0                0.0      0.0   
97                 0.0             0.0                0.0      0.0   
98                 0.0             0.0                0.0      0.0   
99                 0.0             0.0                0.0      0.0   
100                0.0             0.0                0.0      0.0   

     Airport Gate  Airport Lounge  Airport Service  Airport Terminal  ...  \
0             0.0             0.0              0.0               0.0  ...   
1             0.0             0.0              0.0               0.0  ...   
2             0.0             0.0              0.0               0.0  ...   
3             0.0             0.0              0.0               0.0  ...   
4             0.0             0.0              0.0               0.0  ...   
..            ...             ...              ...               ...  ...   
96            0.0             0.0              0.0               0.0  ...   
97            0.0             0.0              0.0               0.0  ...   
98            0.0             0.0              0.0               0.0  ...   
99            0.0             0.0              0.0               0.0  ...   
100           0.0             0.0              0.0               0.0  ...   

     Trail  Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0      0.0            0.0                            0.0               0.0   
1      0.0            0.0                            0.0               0.0   
2      0.0            0.0                            0.0               0.0   
3      0.0            0.0                            0.0               0.0   
4      0.0            0.0                            0.0               0.0   
..     ...            ...                            ...               ...   
96     0.0            0.0                            0.0               0.0   
97     0.0            0.0                            0.0               0.0   
98     0.0            0.0                            0.0               0.0   
99     0.0            0.0                            0.0               0.0   
100    0.0            0.0                            0.0               0.0   

     Vietnamese Restaurant  Warehouse Store  Wine Bar  Wine Shop  Wings Joint  \
0                      0.0              0.0       0.0        0.0          0.0   
1                      0.0              0.0       0.0        0.0          0.0   
2                      0.0              0.0       0.0        0.0          0.0   
3                      0.0              0.0       0.0        0.0          0.0   
4                      0.0              0.0       0.0        0.0          0.0   
..                    

### New size of the data

In [24]:
Toronto_grouped.shape

(101, 257)

### Create a function to get the top most common venue categories

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

### The top 10 venues for each neighborhood

In [26]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = Toronto_grouped['Neighborhood']

for ind in np.arange(Toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(Toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood      1st Most Common Venue  \
0                                        Agincourt  Latin American Restaurant   
1                           Alderwood, Long Branch                Pizza Place   
2  Bathurst Manor, Wilson Heights, Downsview North                Coffee Shop   
3                                  Bayview Village         Chinese Restaurant   
4                Bedford Park, Lawrence Manor East                Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Lounge          Skating Rink        Breakfast Spot   
1                   Gym        Sandwich Place                   Pub   
2                  Bank              Pharmacy      Sushi Restaurant   
3                  Café   Japanese Restaurant                  Bank   
4        Sandwich Place            Restaurant      Sushi Restaurant   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
0         Women's Store      Department Store  Eastern European Restaurant   
1           Coffee Shop              Pharmacy                 Dance Studio   
2            Restaurant     Mobile Phone Shop               Sandwich Place   
3   Distribution Center    Dim Sum Restaurant                        Diner   
4     Indian Restaurant               Butcher                         Café   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0             Drugstore            Donut Shop       Doner Restaurant  
1             Drugstore            Donut Shop       Doner Restaurant  
2           Pizza Place    Chinese Restaurant            Bridal Shop  
3        Discount Store         Women's Store       Department Store  
4  Fast Food Restaurant    Italian Restaurant                    Spa

### Cluster Neighborhoods

In [27]:
# Run k-means to cluster the neighborhood into 5 clusters

# set number of clusters
kclusters = 5

Toronto_grouped_clustering = Toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(Toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3])

### Labeling with K-Means

In [28]:
kmeans.labels_[0:100]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       4, 0, 0, 0, 0, 0, 0, 3, 0, 0, 0, 3, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0,
       3, 0, 0, 3, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 3, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 3, 0, 0, 3])

In [29]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

Toronto_merged = combined_data

# merge Toronto_grouped with Toronto_data to add latitude/longitude for each neighborhood
Toronto_merged = Toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

Toronto_merged.head()

PostalCode           Borough                      Neighborhood   Latitude  \
0        M3A        North York                         Parkwoods  43.753259   
1        M4A        North York                  Victoria Village  43.725882   
2        M5A  Downtown Toronto         Regent Park, Harbourfront  43.654260   
3        M6A        North York  Lawrence Manor, Lawrence Heights  43.718518   
4        M7A      Queen's Park     Ontario Provincial Government  43.662301   

   Longitude  Cluster Labels 1st Most Common Venue  2nd Most Common Venue  \
0 -79.329656             3.0                  Park      Food & Drink Shop   
1 -79.315572             0.0           Coffee Shop  Portuguese Restaurant   
2 -79.360636             0.0           Coffee Shop                   Park   
3 -79.464763             0.0        Clothing Store      Accessories Store   
4 -79.389494             0.0           Coffee Shop       Sushi Restaurant   

  3rd Most Common Venue  4th Most Common Venue 5th Most Common Venue  \
0  Fast Food Restaurant          Women's Store        Discount Store   
1           Pizza Place           Hockey Arena          Dance Studio   
2                   Pub                 Bakery                  Café   
3           Coffee Shop  Vietnamese Restaurant             Gift Shop   
4         Burrito Place            Yoga Studio                  Park   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0      Department Store          Dessert Shop    Dim Sum Restaurant   
1         Deli / Bodega      Department Store          Dessert Shop   
2          Dessert Shop            Beer Store        Farmers Market   
3              Boutique    Miscellaneous Shop   Sporting Goods Shop   
4    Persian Restaurant    Mexican Restaurant                  Café   

  9th Most Common Venue 10th Most Common Venue  
0                 Diner                Dog Run  
1    Dim Sum Restaurant                  Diner  
2                 Hotel                   Bank  
3         Women's Store         Discount Store  
4          Burger Joint            Salad Place

### Removing all the NaN values to prevent data skew

In [30]:
Toronto_merged_nonan = Toronto_merged.dropna(subset=['Cluster Labels'])

In [31]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(Toronto_merged_nonan['Latitude'], Toronto_merged_nonan['Longitude'], Toronto_merged_nonan['Neighborhood'], Toronto_merged_nonan['Cluster Labels']):
    label = folium.Popup('Cluster ' + str(int(cluster) +1) + '\n' + str(poi) , parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[int(cluster-1)],
        fill=True,
        fill_color=rainbow[int(cluster-1)]
        ).add_to(map_clusters)
        
map_clusters

## Let's Examine Clusters

### Cluster-1

In [32]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 0, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
1               North York             0.0           Coffee Shop   
2         Downtown Toronto             0.0           Coffee Shop   
3               North York             0.0        Clothing Store   
4             Queen's Park             0.0           Coffee Shop   
6              Scarborough             0.0  Fast Food Restaurant   
..                     ...             ...                   ...   
96        Downtown Toronto             0.0           Coffee Shop   
97        Downtown Toronto             0.0           Coffee Shop   
99        Downtown Toronto             0.0      Sushi Restaurant   
100  East Toronto Business             0.0    Light Rail Station   
102              Etobicoke             0.0            Kids Store   

     2nd Most Common Venue 3rd Most Common Venue  4th Most Common Venue  \
1    Portuguese Restaurant           Pizza Place           Hockey Arena   
2                     Park                   Pub                 Bakery   
3        Accessories Store           Coffee Shop  Vietnamese Restaurant   
4         Sushi Restaurant         Burrito Place            Yoga Studio   
6            Women's Store  Ethiopian Restaurant      Electronics Store   
..                     ...                   ...                    ...   
96             Pizza Place    Chinese Restaurant                   Café   
97          Sandwich Place                  Café                  Hotel   
99     Japanese Restaurant           Coffee Shop                Gay Bar   
100   Gym / Fitness Center         Garden Center          Burrito Place   
102         Sandwich Place       Supplement Shop            Flower Shop   

           5th Most Common Venue 6th Most Common Venue  \
1                   Dance Studio         Deli / Bodega   
2                           Café          Dessert Shop   
3                      Gift Shop              Boutique   
4                           Park    Persian Restaurant   
6    Eastern European Restaurant             Drugstore   
..                           ...                   ...   
96                    Restaurant              Pharmacy   
97           Japanese Restaurant                  Bank   
99                    Restaurant                   Gym   
100                   Skate Park         Auto Workshop   
102                Tanning Salon        Discount Store   

        7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
1            Department Store          Dessert Shop    Dim Sum Restaurant   
2                  Beer Store        Farmers Market                 Hotel   
3          Miscellaneous Shop   Sporting Goods Shop         Women's Store   
4          Mexican Restaurant                  Café          Burger Joint   
6                  Donut Shop      Doner Restaurant               Dog Run   
..                        ...                   ...                   ...   
96                     Bakery    Italian Restaurant                   Pub   
97                        Gym      Asian Restaurant         Deli / Bodega   
99   Mediterranean Restaurant         Burrito Place     Indian Restaurant   
100      Fast Food Restaurant        Farmers Market                   Spa   
102             Burrito Place          Burger Joint     Convenience Store   

    10th Most Common Venue  
1                    Diner  
2                     Bank  
3           Discount Store  
4              Salad Place  
6      Distribution Center  
..                     ...  
96       Indian Restaurant  
97              Restaurant  
99    Fast Food Restaurant  
100                 Garden  
102                 Bakery  

[87 rows x 12 columns]

### Cluster-2

In [33]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 1, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
62  Central Toronto             1.0                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
62         Women's Store          Dance Studio     Electronics Store   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
62  Eastern European Restaurant             Drugstore            Donut Shop   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
62      Doner Restaurant               Dog Run    Distribution Center

### Cluster-3

In [34]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 2, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
45  North York             2.0             Cafeteria         Women's Store   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
45         Deli / Bodega           Escape Room     Electronics Store   

          6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
45  Eastern European Restaurant             Drugstore            Donut Shop   

   9th Most Common Venue 10th Most Common Venue  
45      Doner Restaurant                Dog Run

### Cluster-4

In [35]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 3, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0               North York             3.0                  Park   
21                    York             3.0                  Park   
35  East York/East Toronto             3.0                  Park   
40              North York             3.0                  Park   
52              North York             3.0                  Park   
61         Central Toronto             3.0                  Park   
66              North York             3.0     Electronics Store   
85             Scarborough             3.0                  Park   
91        Downtown Toronto             3.0                  Park   
98               Etobicoke             3.0                 River   

   2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0      Food & Drink Shop  Fast Food Restaurant                Women's Store   
21         Women's Store                  Pool  Eastern European Restaurant   
35          Intersection     Convenience Store                Women's Store   
40               Airport         Women's Store          Distribution Center   
52         Women's Store          Dance Studio            Electronics Store   
61           Swim School              Bus Line                Deli / Bodega   
66                  Park     Convenience Store                Women's Store   
85          Intersection            Playground                Women's Store   
91                 Trail            Playground                Deli / Bodega   
98                  Park           Curling Ice  Eastern European Restaurant   

          5th Most Common Venue        6th Most Common Venue  \
0                Discount Store             Department Store   
21                    Drugstore                   Donut Shop   
35          Distribution Center                 Dessert Shop   
40             Department Store                 Dessert Shop   
52  Eastern European Restaurant                    Drugstore   
61            Electronics Store  Eastern European Restaurant   
66               Discount Store             Department Store   
85               Discount Store             Department Store   
91             Department Store                 Dessert Shop   
98                    Drugstore                   Donut Shop   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Dessert Shop    Dim Sum Restaurant                 Diner   
21      Doner Restaurant               Dog Run           Curling Ice   
35    Dim Sum Restaurant                 Diner        Discount Store   
40    Dim Sum Restaurant                 Diner        Discount Store   
52            Donut Shop      Doner Restaurant               Dog Run   
61             Drugstore            Donut Shop      Doner Restaurant   
66          Dessert Shop    Dim Sum Restaurant                 Diner   
85          Dessert Shop    Dim Sum Restaurant                 Diner   
91    Dim Sum Restaurant                 Diner        Discount Store   
98      Doner Restaurant               Dog Run   Distribution Center   

   10th Most Common Venue  
0                 Dog Run  
21    Distribution Center  
35                Dog Run  
40       Doner Restaurant  
52    Distribution Center  
61                Dog Run  
66                Dog Run  
85                Dog Run  
91    Distribution Center  
98         Discount Store

### Cluster-5

In [36]:
Toronto_merged_nonan.loc[Toronto_merged_nonan['Cluster Labels'] == 4, Toronto_merged_nonan.columns[[1] + list(range(5, Toronto_merged_nonan.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
57   North York             4.0        Baseball Field         Women's Store   
101   Etobicoke             4.0        Baseball Field         Women's Store   

    3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
57            Event Space           Escape Room     Electronics Store   
101           Event Space           Escape Room     Electronics Store   

           6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
57   Eastern European Restaurant             Drugstore            Donut Shop   
101  Eastern European Restaurant             Drugstore            Donut Shop   

    9th Most Common Venue 10th Most Common Venue  
57       Doner Restaurant                Dog Run  
101      Doner Restaurant                Dog Run